src/sip-call.ts

```diff
@@ -0,0 +1,297 @@
+    lines: string[] = sdp.splitLines(sections[0]),
+    cLine = lines.find((line: string) => line.startsWith('c='))!
+
+  return {
+    address: cLine.match(/c=IN IP4 (\S*)/)![1],
+    audio: getRtpDescription(log, sections, 'audio')
+  }
+}
+
+export class SipCall {
+  private seq = 20
+  private fromParams = { tag: getRandomId() }
+  private toParams: { tag?: string } = {}
+  private callId = getRandomId()
+  private sipStack: SipStack
+  public onEndedByRemote?: () => void
+  private destroyed = false
+  private readonly log: Logger
+
+  constructor(
+    log: Logger,
+    private sipOptions: SipOptions,
+  ) {
+    this.log = log;
+    const fromUri = sip.parseUri(this.sipOptions.from);
+    const host = fromUri.host;
+    const port = fromUri.port;
+
+    const { from } = this.sipOptions
+    this.sipStack = {
+      makeResponse: sip.makeResponse, 
+      ...sip.create({
+        host,
+        hostname: host,
+        port: port,
+        udp: true,
+        tcp: false,
+        tls: false,
+        ws: false
+      },
+      (request: SipRequest) => {
+        if (request.method === 'BYE') {
+          this.log.info('received BYE from remote end')
+          this.sipStack.send(this.sipStack.makeResponse(request, 200, 'Ok'))
+
+          if (this.onEndedByRemote) {
+            this.onEndedByRemote()
+          }
+        } 
+      }
+    )}
+  }
+
+  request({
+    method,
+    headers,
+    content,
+    seq,
+  }: {
+    method: string
+    headers?: Partial<SipHeaders>
+    content?: string
+    seq?: number
+  }) {
+    if (this.destroyed) {
+      return Promise.reject(
+        new Error('SIP request made after call was destroyed')
+      )
+    }
+
+    return new Promise<SipResponse>((resolve, reject) => {
+      seq = seq || this.seq++
+      this.sipStack.send(
+        {
+          method,
+          uri: this.sipOptions.to,
+          headers: {
+            to: {
+              name: '"SIP doorbell client"',
+              uri: this.sipOptions.to,
+              params: this.toParams,
+            },
+            from: {
+              uri: this.sipOptions.from,
+              params: this.fromParams,
+            },
+            'max-forwards': 70,
+            'call-id': this.callId,
+            cseq: { seq, method },
+            ...headers,
+          },
+          content: content || '',
+        },
+        (response: SipResponse) => {
+          if (response.headers.to.params && response.headers.to.params.tag) {
+            this.toParams.tag = response.headers.to.params.tag
+          }
+
+          if (response.status >= 300) {
+            if (response.status !== 408 || method !== 'BYE') {
+              this.log.error(
+                `sip ${method} request failed with status ` + response.status
+              )
+            }
+            reject(
+              new Error(
+                `sip ${method} request failed with status ` + response.status
+              )
+            )
```

Comment:
While debugging my SIP error, I noticed there's an error `reason` that might be useful to log

```suggestion
            if (response.status !== 408 || method !== 'BYE') {
              this.log.error(
                `sip ${method} request failed with status ${response.status} and reason ${response.reason}`
              )
            }
            reject(
              new Error(
                `sip ${method} request failed with status ${response.status} and reason ${response.reason}`
              )
            )
```

Reply:
Have you tried running tcpdump on the OS you're running this solution on? I'd also suggest if your 2N intercom supports  a packet capture to also capture there and download.  If the 2N doesn't support it and its hard wired you could do port mirroring on your network switch and plug another computer into the switch on the port you do the mirror on and do either another tcpdump or just direct into wireshark.  You'll want to put the nic on the computer you plugged into the "Mirrored Port" into promiscuous mode.  Then open both wire captures into wire shark and look for sip registrations and check out the flow charts.  Should give you another view into the issue.

Reply:
Ah good idea, the 2N supports packet capture.

Here's a trace when this plugin makes a SIP call which fails.

```
INVITE sip:192.168.1.3 SIP/2.0
allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, NOTIFY, MESSAGE, SUBSCRIBE, INFO, UPDATE
call-id: 278378
contact: <sip:192.168.1.2>
content-length: 110
content-type: application/sdp
cseq: 22 INVITE
from: <sip:192.168.1.2>;tag=302689
max-forwards: 70
supported: replaces, outbound
to: "SIP doorbell client" <sip:192.168.1.3>;tag
via: SIP/2.0/UDP 192.168.1.2:5060;branch=z9hG4bK373682;rport

v=0
o=192.168.1.2 3747 461 IN IP4 192.168.1.2
s=Talk
c=IN IP4 192.168.1.2
t=0 0
m=audio 15086 RTP/AVP 0
INVITE SIP/2.0
call-id: 278378
content-length: 0
cseq: 22 INVITE
from: <sip:192.168.1.2>;tag=302689
to: "SIP doorbell client" <sip:192.168.1.3>;tag
user-agent: 2N IP Solo 2.35.1.47.3
via: SIP/2.0/UDP 192.168.1.2:5060;branch=z9hG4bK373682;rport=5060


INVITE SIP/2.0
call-id: 278378
content-length: 0
cseq: 22 INVITE
from: <sip:192.168.1.2>;tag=302689
to: "SIP doorbell client" <sip:192.168.1.3>;tag
user-agent: 2N IP Solo 2.35.1.47.3
via: SIP/2.0/UDP 192.168.1.2:5060;branch=z9hG4bK373682;rport=5060


ACK sip:192.168.1.3 SIP/2.0
call-id: 278378
content-length: 0
cseq: 22 ACK
from: <sip:192.168.1.2>;tag=302689
max-forwards: 70
to: "SIP doorbell client" <sip:192.168.1.3>;tag
via: SIP/2.0/UDP 192.168.1.2:5060;branch=z9hG4bK373682;rport


BYE sip:192.168.1.3 SIP/2.0
call-id: 278378
content-length: 0
cseq: 23 BYE
from: <sip:192.168.1.2>;tag=302689
max-forwards: 70
to: "SIP doorbell client" <sip:192.168.1.3>;tag
via: SIP/2.0/UDP 192.168.1.2:5060;branch=z9hG4bK67662;rport


BYE SIP/2.0
call-id: 278378
content-length: 0
cseq: 23 BYE
from: <sip:192.168.1.2>;tag=302689
to: "SIP doorbell client" <sip:192.168.1.3>;tag
user-agent: 2N IP Solo 2.35.1.47.3
via: SIP/2.0/UDP 192.168.1.2:5060;branch=z9hG4bK67662;rport=5060
```

And here's a trace when MicroSIP succeeds

```
INVITE sip:192.168.1.3 SIP/2.0
allow: PRACK, INVITE, ACK, BYE, CANCEL, UPDATE, INFO, SUBSCRIBE, NOTIFY, REFER, MESSAGE, OPTIONS
call-id: 4dd6e58708a84a8ebddc5b0349e428a0
contact: <sip:192.168.1.33:53499;ob>
content-length: 364
content-type: application/sdp
cseq: 24888 INVITE
from: <sip:192.168.1.33>;tag=33533c0e25d246fbb81c310c75d3f6f4
max-forwards: 70
min-se: 90
session-expires: 1800
supported: replaces, 100rel, timer, norefersub
to: <sip:192.168.1.3>
user-agent: MicroSIP/3.21.2
via: SIP/2.0/UDP 192.168.1.33:53499;rport;branch=z9hG4bKPj2f7c6a5736cf442a87dc89a1a29f3b19

v=0
o=- 3871901177 3871901177 IN IP4 192.168.1.33
s=pjmedia
b=AS:84
t=0 0
a=X-nat:0
m=audio 4000 RTP/AVP 9 8 0 101
c=IN IP4 192.168.1.33
b=TIAS:64000
a=rtcp:4001 IN IP4 192.168.1.33
a=sendrecv
a=rtpmap:9 G722/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ssrc:1122779591 cname:4d4a6ffd081c26cc


INVITE SIP/2.0
call-id: 4dd6e58708a84a8ebddc5b0349e428a0
content-length: 0
cseq: 24888 INVITE
from: <sip:192.168.1.33>;tag=33533c0e25d246fbb81c310c75d3f6f4
to: <sip:192.168.1.3>
user-agent: 2N IP Solo 2.35.1.47.3
via: SIP/2.0/UDP 192.168.1.33:53499;rport=53499;branch=z9hG4bKPj2f7c6a5736cf442a87dc89a1a29f3b19


INVITE SIP/2.0
call-id: 4dd6e58708a84a8ebddc5b0349e428a0
contact: <sip:192.168.1.3:5060>
content-length: 0
cseq: 24888 INVITE
from: <sip:192.168.1.33>;tag=33533c0e25d246fbb81c310c75d3f6f4
to: <sip:192.168.1.3>;tag=1119705459
user-agent: 2N IP Solo 2.35.1.47.3
via: SIP/2.0/UDP 192.168.1.33:53499;rport=53499;branch=z9hG4bKPj2f7c6a5736cf442a87dc89a1a29f3b19


INVITE SIP/2.0
allow: REGISTER, INVITE, ACK, CANCEL, OPTIONS, BYE, INFO, NOTIFY
call-id: 4dd6e58708a84a8ebddc5b0349e428a0
contact: <sip:192.168.1.3:5060>
content-length: 232
content-type: application/sdp
cseq: 24888 INVITE
from: <sip:192.168.1.33>;tag=33533c0e25d246fbb81c310c75d3f6f4
to: <sip:192.168.1.3>;tag=1119705459
user-agent: 2N IP Solo 2.35.1.47.3
via: SIP/2.0/UDP 192.168.1.33:53499;rport=53499;branch=z9hG4bKPj2f7c6a5736cf442a87dc89a1a29f3b19

v=0
o=- 1469393603 2233767474801975965 IN IP4 192.168.1.3
s=HIP 2.35.1.47.3
c=IN IP4 192.168.1.3
t=0 0
m=audio 9000 RTP/AVP 9 101
c=IN IP4 192.168.1.3
a=rtpmap:9 G722/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16


ACK sip:192.168.1.3:5060 SIP/2.0
call-id: 4dd6e58708a84a8ebddc5b0349e428a0
content-length: 0
cseq: 24888 ACK
from: <sip:192.168.1.33>;tag=33533c0e25d246fbb81c310c75d3f6f4
max-forwards: 70
to: <sip:192.168.1.3>;tag=1119705459
via: SIP/2.0/UDP 192.168.1.33:53499;rport;branch=z9hG4bKPj8395320e8cda4a5ab740b331c87396db


BYE sip:192.168.1.3:5060 SIP/2.0
call-id: 4dd6e58708a84a8ebddc5b0349e428a0
content-length: 0
cseq: 24889 BYE
from: <sip:192.168.1.33>;tag=33533c0e25d246fbb81c310c75d3f6f4
max-forwards: 70
to: <sip:192.168.1.3>;tag=1119705459
user-agent: MicroSIP/3.21.2
via: SIP/2.0/UDP 192.168.1.33:53499;rport;branch=z9hG4bKPj0f16e470e636447e8c592da9c6c777fd


BYE SIP/2.0
call-id: 4dd6e58708a84a8ebddc5b0349e428a0
content-length: 0
cseq: 24889 BYE
from: <sip:192.168.1.33>;tag=33533c0e25d246fbb81c310c75d3f6f4
to: <sip:192.168.1.3>;tag=1119705459
user-agent: 2N IP Solo 2.35.1.47.3
via: SIP/2.0/UDP 192.168.1.33:53499;rport=53499;branch=z9hG4bKPj0f16e470e636447e8c592da9c6c777fd
```

Reply:
Have you verified the port is open on the homebridge server running the plugin?  I believe the SIP service is called baresip.  Make sure in logs something else isn't attempting to attach to same port.  I believe you've said you used the desktop sip app to test the camera but I didn't see anything about testing the plugin as well.  Make sure your firewall is open.  run this from a remote machine to make sure you can hit the box external.  You can change the range from 1-65535 to be smaller just to cover your needed scope.   

This command will show you what ports services are listening on.
sudo netstat -tulpn | grep LISTEN

This ran from remote machine toward the homebridge will let you know if you firewall or something else is stopping inward connection attempts.
nc -z -v <IP-ADDRESS> 1-65535 2>&1 | grep -v 'Connection refused'

Reply:
Turns out it wasn't an issue with the port.

Through adding a whole bunch of logging and trial and error, I narrowed it down to the `to` header having an empty tag in the INVITE request.

For example, this plugin was sending an INVITE request with the `to` header as

```to: "SIP doorbell client" <sip:192.168.1.3>;tag```

Whereas MicroSIP was sending

```to: <sip:192.168.1.3>```

If I modify https://github.com/Sunoo/homebridge-camera-ffmpeg/pull/1253/files#diff-9195b8330fc9c58adfffb0f687ffa35577548a16e2960e06b91f29c20db2be1eR181 to
```
            params: this.toParams.tag ? this.toParams : undefined,
```

so it does not send an empty `tag` if it is `undefined`, then the INVITE chain all works correctly. I see a `ringing` and `OK` response afterwards.

Now I'm running into a separate issue where my FFmpeg is erroring starting up.

```
[12/09/2022, 8:52:44 pm] [Camera FFmpeg] [Intercom test] Starting video stream: 1280 x 720, 30 fps, 299 kbps (AAC-eld)
[12/09/2022, 8:52:44 pm] [Camera FFmpeg] [Intercom test] FFmpeg exited with code: 1 and signal: null (Error)
```

I presume it's erroring trying to initialize the SIP call. I didn't have enough time today to figure out why but will aim to do more debugging soon.

Reply:
As I recall you can find the capture string that gets pushed into FFmpeg and try and manually run it on cli  or just run a manual local instance to make sure FFmpeg is working correctly.

Here are some old ffmpeg test streams I have for examples. I don't think its a sip related error but could be wrong. According to info written by developer the audio stream is completely seperate from the video stream see snippet below.

"It adds support for two-way audio between Homekit and a SIP-based video doorbell.
Those video doorbells based on SIP seem to exist in at least two different flavours:

1. Video is a completely seperate thing: e.g. MJPEG stream via HTTP only. SIP is only used for Audio.
2. Video is a H264 stream that is also negotiated as an additional stream (next to Audio) during SIP signaling.
This PR only addresses variant 1. However, adding variant 2. should not be too difficult."

Some sample ffmpeg test streams just modify to your needs to test your own ffmpeg install. Just change IP, Username, password and stream channel ID's. The bottom is probably the simplest.  You can use find the stream URL using the Windows ONVIF Device Manager tool.

Just FYI the URL's below are from different testing I was doing. Not Vid / SIP.  His documentation above says it does HTTP top example not RTSP other 2 examples. If my info is not being helpful feel free to tell me thanks but no thanks. 

ffmpeg -i http://server:port -c copy -f mpegts somefile.ts

ffmpeg -rtsp_transport tcp -allowed_media_types video -i "rtsp://user:password@192.168.0.65:554/0" -c copy -f h264 - | ffmpeg -f h264 -r 30 -i - -f lavfi -i anullsrc -b:a 2k -c:a aac -c:v copy -f flv "rtmp://a.rtmp.youtube.com/live2/YOUR_STREAM_KEY_HERE"

ffmpeg -rtsp_transport tcp -re -i rtsp://user:password@10.0.29.230/Streaming/Channels/102 test.mp4

Reply:
Figured out the issue with the FFmpeg stream error which led to a bug https://github.com/Sunoo/homebridge-camera-ffmpeg/pull/1253/files#r969484626

However even after fixing this I'm still not getting the return audio working. Will need to debug further. It feels good chipping away at this every day after work.

Reply:
I've been trying to debug why the "two-way" FFmpeg stream keeps crashing. I added `debugReturn: true` to my config and saw this in the logs

```[14/09/2022, 8:44:48 pm] [Camera FFmpeg] [Intercom test] Starting video stream: 1280 x 720, 30 fps, 299 kbps (AAC-eld)
[14/09/2022, 8:44:48 pm] [Camera FFmpeg] [Intercom test] [Two-way] Stream command: C:\Users\Long\Desktop\homebridge-camera-ffmpeg-sipdoorbell\node_modules\ffmpeg-for-homebridge\ffmpeg.exe -hide_banner -protocol_whitelist pipe,udp,rtp,file,crypto -f sdp -c:a libfdk_aac -i pipe:  -acodec pcm_mulaw -ac 1 -ar 8k -f rtp -payload_type 0 rtp://127.0.0.1:15128?rtcpport=15129 -loglevel level+verbose
[14/09/2022, 8:44:48 pm] [Camera FFmpeg] [Intercom test] [Two-way] [sdp @ 000001bad9fe6980] [verbose] setting jitter buffer size to 500
[14/09/2022, 8:44:59 pm] [Camera FFmpeg] [Intercom test] [Two-way] [warning] Guessed Channel Layout for Input Stream #0.0 : mono
[14/09/2022, 8:44:59 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info] Input #0, sdp, from 'pipe:':
[14/09/2022, 8:44:59 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]   Metadata:
[14/09/2022, 8:44:59 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]     title           : Talk
[14/09/2022, 8:44:59 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]   Duration: N/A, bitrate: N/A
[14/09/2022, 8:44:59 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]     Stream #0:0: Audio: aac, 16000 Hz, mono, s16
[14/09/2022, 8:44:59 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info] Stream mapping:
[14/09/2022, 8:44:59 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]   Stream #0:0 -> #0:0 (aac (libfdk_aac) -> pcm_mulaw (native))
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [error] pipe:: Unknown error
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [graph_0_in_0_0 @ 000001bada5aa3c0] [verbose] tb:1/16000 samplefmt:s16 samplerate:16000 chlayout:0x4
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [format_out_0_0 @ 000001bada5ac800] [verbose] auto-inserting filter 'auto_resampler_0' between the filter 'Parsed_anull_0' and the filter 'format_out_0_0'
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [auto_resampler_0 @ 000001bada5ace00] [verbose] ch:1 chl:mono fmt:s16 r:16000Hz -> ch:1 chl:mono fmt:s16 r:8000Hz
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info] Output #0, rtp, to 'rtp://127.0.0.1:15128?rtcpport=15129':
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]   Metadata:
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]     title           : Talk
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]     encoder         : Lavf58.38.101
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]     Stream #0:0: Audio: pcm_mulaw, 8000 Hz, mono, s16, 64 kb/s
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]     Metadata:
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info]       encoder         : Lavc58.70.100 pcm_mulaw
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info] size=       0kB time=00:00:00.00 bitrate=N/A speed=   0x    
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [verbose] No more output streams to write to, finishing.
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info] size=       0kB time=00:00:00.00 bitrate=N/A speed=   0x    
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [info] video:0kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [verbose] Input file #0 (pipe:):
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [verbose]   Input stream #0:0 (audio): 0 packets read (0 bytes); 0 frames decoded (0 samples); 
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [verbose]   Total: 0 packets (0 bytes) demuxed
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [verbose] Output file #0 (rtp://127.0.0.1:15128?rtcpport=15129):
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [verbose]   Output stream #0:0 (audio): 0 frames encoded (0 samples); 0 packets muxed (0 bytes); 
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [verbose]   Total: 0 packets (0 bytes) muxed
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [warning] Output file is empty, nothing was encoded (check -ss / -t / -frames parameters if used)
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [AVIOContext @ 000001bad9feaf40] [verbose] Statistics: 0 seeks, 0 writeouts
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] [AVIOContext @ 000001bad9fe7400] [verbose] Statistics: 356 bytes read, 0 seeks
[14/09/2022, 8:45:10 pm] [Camera FFmpeg] [Intercom test] [Two-way] FFmpeg exited with code: 0 and signal: null (Error)
```

The notable error seem to be `[Two-way] [error] pipe:: Unknown error` which seems to imply something's wrong with the pipe input which I dumped as 

```
v=0
o=- 0 0 IN IP4 192.168.1.141
s=Talk
c=IN IP4 192.168.1.141
t=0 0
m=audio 14695 RTP/AVP 110
b=AS:24
a=rtpmap:110 MPEG4-GENERIC/16000/1
a=rtcp-mux
a=fmtp:110 profile-level-id=1;mode=AAC-hbr;sizelength=13;indexlength=3;indexdeltalength=3; config=F8F0212C00BC00
a=crypto:1 AES_CM_128_HMAC_SHA1_80 inline:kPiXERZJq/CUIDqyz54SFAWKDav7MevvzenVBd4G
```

Reply:
I've been debugging this more and am still stuck. I've tried changing to call a SIP client on my PC and running into the same issue with the two-way FFmpeg process erroring with `Unknown error`.

I've also tried changing from a pipe input to writing the SDP data into a file and loading that, same error, so I don't think it's an issue with the pipe but the stream directly.

@nanosonde not sure if you have any ideas about why the two-way input stream might be erroring like this?

Reply:
Figured out the problem was with Windows Firewall. Because I was testing this plugin in a different directory using `npm link`, the `ffmpeg.exe` process wasn't allowed in the firewall to receive incoming connections which the RTP connection was. After I added it to the allow list the stream works.

Reply:
Good catch. I was running on the assumption you were running Linux. To many unknowns / minimal support with Microsoft. You're braver than I am.